### PR TITLE
[JENKINS-33389] Adapt to new parent POM and split JTH

### PR DESCRIPTION
**Pull request by andresrc**
[JENKINS-33389](https://issues.jenkins-ci.org/browse/JENKINS-33389)

* [x] Adapt to new parent POM
* [x] Adapt to split JTH.
* [x] Fix javadocs for Java 8
* [x] Correct line breaks in modified files
* [x] Migrate tests from `HudsonTestCase` to `JenkinsRule`.
* [x] Check PCT scenario against 1.642.3

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -22,7 +22,11 @@
     </licenses>
 
     <properties>
-        <workflow.version>1.1</workflow.version>
+      <jenkins.version>1.580.1</jenkins.version>
+      <!-- TODO check if we'd better add the PI. -->
+      <jelly.requirePI>false</jelly.requirePI>
```
Comment (jglick) [inline pom.xml]: Yes you should.

**Comment by andresrc** [comment]
CI builds crashing (one was OOM, the other did not capture forked VM error). Tested locally.

**Comment by ikedam** [comment]
JTH = Jenkins Test Harness

See also:
* [PROPOSAL] New Parent POM for Jenkins Plugins
    * https://groups.google.com/forum/#!msg/jenkinsci-dev/3_IMnXmtuDw/TEeowLK-BgAJ
    * https://github.com/jenkinsci/plugin-pom
* [PROPOSAL] Split of `jenkins-test-harness` out of `jenkinsci/jenkins`
    * https://groups.google.com/forum/#!msg/jenkinsci-dev/zviuh6z5_KM/OUKrvgvGAwAJ
    * https://github.com/jenkinsci/jenkins-test-harness
    * https://github.com/jenkinsci/jenkins-test-harness-tools

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java [hunk 2]
```diff
@@ -73,14 +73,7 @@
...
 import org.junit.Test;
-import org.jvnet.hudson.test.Bug;
-import org.jvnet.hudson.test.ExtractResourceSCM;
-import org.jvnet.hudson.test.HudsonTestCase;
-import org.jvnet.hudson.test.Issue;
-import org.jvnet.hudson.test.CaptureEnvironmentBuilder;
-import org.jvnet.hudson.test.FailureBuilder;
-import org.jvnet.hudson.test.MockFolder;
-import org.jvnet.hudson.test.UnstableBuilder;
+import org.jvnet.hudson.test.*;
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: I prefer specific imports.

**Comment by ikedam** [comment]
Please turn off autoformatter of your IDE and don't make unrelated whitespace changes.
(even though there're inconsistent formattings or useless whitespaces)

**Comment by ikedam** [comment]
Hmm... whitespace changes in `CopyartifactPermissionProperty.java` looks caused for existing inconsistent line breaks (LFs and CRLFs).
Those changes are acceptable as all editors cannot handle mixed line break characters in a file.

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java (hunk 2 above)
Comment (andresrc) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: So do I, IDE misconfiguration in this project, will fix.

**Comment by andresrc** [comment]
Yes, IDEA proposed to unify line breaks. I did not include it so as not to include unrelated changes. If it's acceptable I will include the changes.

**Comment by andresrc** [comment]
Updated by force-pushing as it was easier to fix the line-breaks in the affected files.

File: pom.xml [hunk 3]
```diff
@@ -64,6 +72,24 @@
         <version>1.3</version>
       </dependency>
       <dependency>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>junit</artifactId>
+        <version>1.10</version>
+        <scope>test</scope>
+      </dependency>
```
Comment (ikedam) [inline pom.xml]: (Note for me)
matrix-project depends on junit-plugin (via Jenkins core).
Jenkins core 1.580.1 bundles junit-1.0.

File: pom.xml [hunk 4]
```diff
@@ -64,6 +72,24 @@
...
+      <dependency>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>credentials</artifactId>
+        <version>1.24</version>
+        <scope>test</scope>
+      </dependency>
```
Comment (ikedam) [inline pom.xml]: (Note for me)
credentials-plugin is not necessary to have tests succeeds.
Though, it is useful to suppress error messages from folder-plugin.

File: pom.xml [hunk 5]
```diff
@@ -64,6 +72,24 @@
...
+      <dependency>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>matrix-auth</artifactId>
+        <version>1.2</version>
+        <scope>test</scope>
+      </dependency>
```
Comment (ikedam) [inline pom.xml]: (Note for me)
Some test codes depends on matrix-auth to configure authorizations.

**Comment by ikedam** [comment]
`mvn test` also fails in my environment
* `CopyArtifactTest` fails with `OutOfMemory`
    * I had to add a following code to pom.xml to run tests.

        ```
  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-surefire-plugin</artifactId>
        <configuration>
            <argLine>-Xmx1024m</argLine>
        </configuration>
      </plugin>
    </plugins>
  </build>
    ```
    * Jenkins-test-hanrness might have a memory leaking issue.
    * We might have to have `CopyArtifactTest` migrate to `JenkinsRule` (junit4)
* `PluginAutomaticTestBuilder$OtherTests#testPluginActive` failed with "Plugin copyartifact failed to start"
    * I have to see more details about that.

**Comment by jglick** [comment]
I switched the build to run on Java 8, let us see if that helps.

File: pom.xml (hunk 5 above)
Comment (jglick) [inline pom.xml]: You can use `MockAuthorizationStrategy` if you pick up a sufficiently new `jenkins-test-harness`.

File: src/main/java/hudson/plugins/copyartifact/BuildSelectorParameter.java [hunk 6]
```diff
@@ -84,7 +85,10 @@ private StringParameterValue toStringValue(BuildSelector selector) {
 
     /**
      * Convert xml fragment into a BuildSelector object.
-     * @throws XStreamException or ClassCastException if input is invalid
+     * @param xml XML fragment to parse.
+     * @return the BuildSelector represented by the input XML.
+     * @throws XStreamException if the object cannot be deserialized
+     * @throws ClassCastException if input is invalid
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/BuildSelectorParameter.java]: You know you need not correct every Javadoc _warning_, right? Just _errors_.

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactPermissionPropertyTest.java [hunk 7]
```diff
@@ -269,4 +270,15 @@ public void testDescriptorDoAutoCompleteProjectNames() throws Exception {
...
     }
+
+    /**
+     * Creates an empty Matrix project with the provided name.
+     *
+     * @param name Project name.
+     * @return an empty Matrix project with the provided name.
+     */
+    private MatrixProject createMatrixProject(String name) throws IOException {
+        return j.jenkins.createProject(MatrixProject.class, name);
```
Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactPermissionPropertyTest.java]: Clearer just to inline IMO.

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java [hunk 8]
```diff
@@ -926,4 +924,16 @@ public void testMavenModule() throws Exception {
...
         );
     }
+
+    /**
+     * Creates an empty Maven project with an unique name.
+     *
+     * @return an empty Maven project with an unique name.
+     */
+    private MavenModuleSet createMavenProject() throws IOException {
+        MavenModuleSet mavenModuleSet = j.jenkins.createProject(MavenModuleSet.class, "test"+j.jenkins.getItems().size());
```
Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: The autogeneration of project names is pretty stupid when every test case gets its own fresh Jenkins instance anyway. Simpler and clearer to just specify a name.

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java [hunk 9]
```diff
@@ -926,4 +924,16 @@ public void testMavenModule() throws Exception {
...
+        mavenModuleSet.setRunHeadless(true);
```
Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: Silly to be repeating this incantation in every test using `maven-plugin`. Better to just fix that plugin to set `runHeadless` to true by default when `Main.isUnitTest()`.

Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: (or always)

**Comment by jglick** [comment]
:bee:

**Comment by ikedam** [comment]
Sorry for leaving this request for so long time... I can't have time to work for this (e.g. migrating CopyArtifactTest to junit4) yet.

@jglick 
Would you tell me why migrating to Java8 fixes `OutOfMomoryError`? (if you know)

**Comment by ikedam** [comment]
Thanks for migrating tests to `JenkinsRule`.
Unfortunately, `OutOfMemoryError` still occurs when tested with Java7.
But I finally found it an existing potential issue (that is, this change doesn't cause that problem):

* Every time a test case run, the usage of PermGen increases.
    * I suppose PermGen is consumed when a plugin is loaded and a new class loader is created.
    * This occurs even with jenkins-test-harness provided by Jenkins 1.580.1.
* The initial PermGen usage of jenkins-test-harness >= 2.0 is larger than that of jenkins-test-harness < 2.0.
* That results `OutOfMemoryError` with jenkins-test-harness >= 2.0. Though, I believe I can trigger the same issue even with jenkins-test-harness < 2.0 just by adding more test cases.
    * `OutOfMemoryError` occurs only when running `mvn test`. `mvn test -Dtest=CopyartifactTest` doesn't cause `OutOfMemoryError` as "enough" test cases run only with `CopyArtifactTest`.
* `OutOfMemoryError` doesn't occur with Java8, as Java8 no longer uses PermGen.

I conclude that this change is good and ready to merge.
I plan to merge this change after other pull requests (#79, #81) merged as that may cause conflicts and requires manual merging.

File: src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java [hunk 10]
```diff
@@ -46,11 +47,8 @@
 
 import org.junit.Rule;
 import org.junit.Test;
-import org.jvnet.hudson.test.Bug;
-import org.jvnet.hudson.test.ExtractResourceSCM;
-import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.*;
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/TriggeredBuildSelectorTest.java]: You'd better to configure your IDE not to use wildcard imports. It's often considered a bad style.

**Comment by andresrc** [comment]
@ikedam Please notify when https://github.com/jenkinsci/copyartifact-plugin/pull/79 and https://github.com/jenkinsci/copyartifact-plugin/pull/81 are merged and I will rebase.

Thanks

**Comment by ikedam** [comment]
#81 caused conflicts but merged manually.
Thanks!

**Comment by ikedam** [comment]
Oops, I had to fix javadocs for changes of parameters in #81.
That caused Jenkins fail after merging.
Let me have time to fix that.

**Pull request merged**